rust/candid/src/parser/random.rs

```diff
@@ -216,3 +121,75 @@ impl IDLValue {
         })
     }
 }
+
+impl IDLArgs {
+    pub fn any(u: &mut Unstructured, env: &TypeEnv, types: &[Type]) -> Result<Self> {
+        let mut args = Vec::new();
+        for t in types.iter() {
+            let mut config = GenConfig::default();
+            let v = config.any(u, env, t)?;
+            args.push(v);
+        }
+        Ok(IDLArgs { args })
+    }
+}
+
+impl TypeEnv {
+    /// Upper bound of type size. Returns None if infinite.
+    fn size_helper(&self, seen: &mut HashSet<String>, t: &Type) -> Option<usize> {
+        use Type::*;
+        Some(match t {
+            Var(id) => {
+                if seen.insert(id.to_string()) {
+                    let ty = self.rec_find_type(id).unwrap();
+                    self.size_helper(seen, &ty)?
+                } else {
+                    return None;
+                }
+            }
+            Empty => 0,
+            Opt(t) => 1 + self.size_helper(seen, t)?,
+            Vec(t) => 1 + self.size_helper(seen, t)? * 2,
+            Record(fs) => {
+                let mut sum = 0;
+                for Field { ty, .. } in fs.iter() {
+                    sum += self.size_helper(seen, ty)?;
+                }
+                1 + sum
+            }
```

Comment:
Technically, the size of a product is, well, the product of the sizes (not the sum). And then, if you want to be correct, you have to make sure that the size of `{empty; int}` comes out as  0, not ∞ (i.e. `0 * ∞ = ∞`, which is not the semantics you get with Rust’s `?` operator here).

Reply:
It's computing the size of the `IDLValue` of type t, so sum is more appropriate. For empty record, it's a bit tricky, because `type A = record { A }` has ∞ value size, but it's actually 0.

Reply:
I don’t follow. Is this the size as in “number of elements” or “size of nodes in the type expressions”?

Surely `{opt bool; opt bool}` has _(1+2)*(1+2) = 9_ different values, not 6.

Reply:
Ah, indeed, size of `bool` is 1, not 2. Ignore me then, _thought_ this is the cardinality of the type (which I would expect to use as a weight when trying to pick elements at random) not the size of the type expression… but maybe using the type expressoin is also sensible.

Reply:
I think all metrics make sense for picking variants: cardinality, depth, value size. But we also use this to estimate the length of a vector, value size feels more accurate.

Reply:
> I think all metrics make sense for picking variants: cardinality, depth, value size. But we also use this to estimate the length of a vector, value size feels more accurate.

I wonder, when is this (very cool) avenue going to fully devolve into fractal geometry, rather than merely flirt with it?

Reply:
E.g., We can also talk of "area" and "dimension" of the values that we get, a la fractal geometry.